### ISSUE TITLE
`/usr/lib/sysimage/libdnf5/nevras.toml` empty after `<ctrl-c>`

### ISSUE THREAD

**Issue by kontura** [rpm-software-management/dnf5#1753]
I was running: `while true; do sudo dnf5 install -y ./htop-3.3.0-1.fc39.x86_64.rpm; sudo dnf5 remove htop -y; done` to debug a different issue and after hitting `<ctrl-c>`  `/usr/lib/sysimage/libdnf5/nevras.toml` ended up empty.
Following runs resulting in:

```
Loading system state TOML file /usr/lib/sysimage/libdnf5/nevras.toml failed (see dnf5-system-state(7)): [error] key "version" not found in the top-level table
 --> /usr/lib/sysimage/libdnf5/nevras.toml
   |
 1 |
   |  the parsed file is empty
```

It likely requires "lucky" timing.

**Comment by ppisar** [comment]
The file is written in State::save() by this code:

~~~~
utils::fs::File(get_nevra_state_path(), "w").write(toml_format(make_top_value("nevras", nevra_states)));
~~~~

The flaw is in libdnf5::utils::fs::File class. It directly writes to the named file instead of writing to a temporary file and then atomically renaming it to the given name.

**Comment by ppisar** [comment]
Atomically replacing a file might not be enough: State::save() writes into many files. An interruption could leave half of the files fully updated and the other half fully outdated. One would have to atomically replace whole directory (which is not portably supported by all file systems), or preserve old new new generation of the files in parallel so that next-excuted DNF can rollback when it discovers an unfinished update.

**Comment by ppisar** [comment]
Close this issue as a duplicate of an older issue #1609.
